src/v/cluster/decommissioning_monitor.cc

```diff
@@ -0,0 +1,256 @@
+    start_decommission_loop();
+    return ss::now();
+}
+
+void decommissioning_monitor::start_decommission_loop() {
+    return (void)ss::with_gate(_bg, [this] {
+        return ss::do_until(
+          [this] { return _as.local().abort_requested(); },
+          [this] {
+              return _new_changes.wait()
+                .then([this] {
+                    return ss::parallel_for_each(
+                      _reallocations, [this](reallocation_meta& meta) {
+                          return reallocate_replica_set(meta);
+                      });
+                })
+                .then([this] {
+                    // remove finished meta
+                    auto it = std::remove_if(
+                      std::begin(_reallocations),
+                      std::end(_reallocations),
+                      [](reallocation_meta& meta) { return meta.is_finished; });
+
+                    _reallocations.erase(it, std::end(_reallocations));
+                })
+                .then([this] {
+                    return _allocator.invoke_on(
+                      partition_allocator::shard,
+                      [](partition_allocator& allocator) {
+                          std::vector<model::node_id> ids;
+                          for (const auto& [id, node] :
+                               allocator.allocation_nodes()) {
+                              if (node->is_decommissioned() && node->empty()) {
+                                  ids.push_back(id);
+                              }
+                          }
+                          return ids;
+                      });
+                })
+                .then([this](std::vector<model::node_id> to_remove) {
+                    // no nodes to reamove from cluster
+                    if (to_remove.empty() || !_raft0->is_leader()) {
+                        return ss::now();
+                    }
+                    vlog(
+                      clusterlog.info,
+                      "removing decommissioned nodes: {}",
+                      to_remove);
+                    // we are not interested in the result as we check
+                    // configuration in next step
+                    return _raft0
+                      ->remove_members(to_remove, model::revision_id{0})
+                      .discard_result();
+                })
+                .then([this] {
+                    auto cfg = _raft0->config();
+                    if (cfg.type() != raft::configuration_type::simple) {
+                        return;
+                    }
+                    // check results
+                    auto it = std::remove_if(
+                      std::begin(_decommissioned),
+                      std::end(_decommissioned),
+                      [&cfg](model::node_id id) {
+                          return !cfg.contains_broker(id);
+                      });
+
+                    _decommissioned.erase(it, std::end(_decommissioned));
```

Comment:
could this whole block be replaced by `std::erase_if` (and a few continuation back that use remove_if)?